### PR TITLE
Issue 13: QR Codes

### LINKED ISSUE #13: Add support for QR code generation
Use case: as a meeting organizer, i want to quickly create an event on my phone, then share that meeting with my friends though a QR code on my phone

QR code would either be the link to the event or the link to the calendar file

### PR DESCRIPTION
**Pull request by Nabushika**
Fixes #13 
## Changes
* Uses `qrcode-generator` crate.
* Cross-service (local/shuttle) router creation, to avoid updates in two places.
* Moved calendar generation to new file `cal.rs` to allow reuse between qr generation and normal function.
* Fixed clippy complaining about a `match` which could have been an `if let`.

### Decisions
* Exports images as PNG for better compatibility (HTML image embedding didn't work with SVG), plus smaller file sizes (<2KB, compared to ~200KB for SVG).
* QR generation is handled by a new endpoint `/qr`, for continued ease of use with `curl` etc., plus this allows people to share links to images rather than the image itself if desired.
* QR contains embedded calendar event, rather than a link back to the app to generate the event.
* QR code has medium ECC - not personally sure if this is correct, we could go lower for smaller codes.
### UI/UX
QR code is tucked beneath the link to the calendar invite:
image

QR code directly produces a calendar event on both Android and iOS:
Android | iOS
-----------|-------
image|image

### CODE REVIEW

**Comment by mre** [comment]
Wow, this is a nice contribution. Thanks a lot @Nabushika.
I tested it locally and it works like a charm. As expected it did create a calendar event on Android.
It's also great that you refactored the router code. Something that I wanted to do as well. 👍 

The QR code has quite a high resolution. I was wondering if e.g. a version 1 or version 2 code would suffice as well.
In any case I think we can tweak that later so it's not a blocker. Just wanted to pick your brain on this before merging it in.

Apart from that the code looks great and the QR code crate is well-maintained, so this is ready to be deployed from my side.

**Comment by Nabushika** [comment]
Having a look see if I can clean up the error handling

**Comment by mre** [comment]
Good job! Shall I merge and deploy it or would you like to make any other changes before?

**Comment by Nabushika** [comment]
As I said, I think I can clean up the error handling a little first, I shouldn't have introduced a new error type since you're using anyhow anyway. If you give me another day that should be enough to clean it up a little?

**Comment by mre** [comment]
Oh, sorry. I misread your message. That's totally fine. 👍

**Comment by Nabushika** [comment]
I cleaned up the start/end code in `create_calendar` and added some tests to make sure it works correctly - I haven't tested the "now" functionality, was a little unsure how to go about that without creating flaky tests, maybe we'd want to split the calendar function up into two functions, one that takes params plus a "now" parameter, and one that just passes in `Utc::now` as the now?
I also got rid of my custom error type, now preferring to use `anyhow::error` like the rest of your code.

Anyway, I think I'm happy with this now. Let me know if you'd like any other changes, otherwise feel free to merge :)

**Comment by Nabushika** [comment]
As for the "high resolution" comment, I believe that's because the whole calendar file is encoded into the QR code, if we made it link to a URL it'd probably be smaller (it grows to fit as you type in more data currently) - but then the user would have to head to the website and maybe download the calendar file before being able to add to their calendar. I think this way has a nicer experience as it comes up directly on the user's phone.

**Comment by Nabushika** [comment]
Okay, actually done now, I promise 😂

File: src/cal.rs [hunk 1]
```diff
@@ -43,19 +43,16 @@ pub fn create_calendar(params: HashMap<String, String>) -> Result<Calendar> {
         .map_or(Ok(None), |r| r.map(Some))
         .map_err(|e| e.context("Invalid duration"))?;
 
-    let (start, end) = match (start, end, duration) {
+    let (start, end) = match (start, end, duration.unwrap_or(DEFAULT_EVENT_DURATION)) {
```
Comment (mre) [inline src/cal.rs:46]: Pretty smart to combine all cases into a `match` statement by the way. 👍

**Comment by mre** [comment]
> Having said that, currently ECC is set to medium, meaning 15% of the qr code can be misread. We could change this to low, which drops it to 7% but would make it smaller.

We can always change it later. I think it's fine actually.
I have to say it's kinda fun to change the data and watch the QR code change. 😆 

> I haven't tested the "now" functionality, was a little unsure how to go about that without creating flaky tests, maybe we'd want to split the calendar function up into two functions, one that takes params plus a "now" parameter, and one that just passes in Utc::now as the now?

Good idea. Maybe that's something we could tackle in the future.

> Okay, actually done now, I promise 😂

Haha, no worries. I'll take your word for it. 😉 Merging and deploying.

**Comment by mre** [comment]
Oookay the deployment failed with some weird caching issue.
Deleting and recreating the project didn't help.
I asked for help in the Shuttle Discord. Please hold on. ^^

**Comment by Nabushika** [comment]
Oh dear

**Pull request merged**